_studio/mfx_lib/shared/src/libmfxsw_encode.cpp

```diff
@@ -54,13 +57,153 @@
 #include "mfx_vp9_encode_hw.h"
 #endif
 
-#if defined (MFX_ENABLE_HEVC_VIDEO_FEI_ENCODE)
 #if defined(MFX_VA)
 #include "mfx_h265_fei_encode_hw.h"
 #include <libmfx_core_interface.h>
 #endif
+
+
+struct CodecKey {
```

Comment:
I'd rather move that custom [CodecKey] logic into a handler u're binding:
```
/*helper function to check if FEI is supported*/
inline
bool check_fei(mfxSession s, mfxU32 /*codec*/)
{
    auto enabled = static_cast<bool*>(session->m_pCORE->QueryCoreInterface(MFXIFEIEnabled_GUID)); 
    return enabled && *enabled;
}

//... codec2helpers table
[](mfxSession s, mfxVideoParam *in, mfxVideoParam *out) 
{
    return check_fei(s, out->mfx.CodecId) ?
        MfxHwH265FeiEncode::H265FeiEncode_HW::Query(s->m_pCORE.get(), in, out) :
        MfxHwH265Encode::MFXVideoENCODEH265_HW::Query(s->m_pCORE.get(), in, out)
    ;
}
```

Profit: 
1. U don't need to introduce a new class to manage FEI handlers, actually it's just one.
2. U don't need to have 2 handlers for each codec.

Reply:
That would be a good idea if we wouldn't have _different_ macro definitions covering FEI and non-FEI. As a result we will once more time have a nightmare disaster. Right now we at lease have straight forward work with the macroses.

Reply:
Isn't this an option?
```
[](mfxSession s, mfxVideoParam *in, mfxVideoParam *out) 
{
    return check_fei(s, out->mfx.CodecId) ?
#if defined (MFX_ENABLE_HEVC_VIDEO_FEI_ENCODE) 
        MfxHwH265FeiEncode::H265FeiEncode_HW::Query(s->m_pCORE.get(), in, out) :
#else
        MFX_ERR_UNDEFINED_BEHAVIOR
#endif
        MfxHwH265Encode::MFXVideoENCODEH265_HW::Query(s->m_pCORE.get(), in, out)
    ;
}
```

I really don't think that having dozen of stubs like
```
{
    MFX_CODEC_AVC, 
    // .fei = 
    false 
}
```
is more suitable.